libs/partners/astradb/langchain_astradb/vectorstores/astradb.py

```diff
@@ -284,6 +372,82 @@ def _get_embedding_dimension(self) -> int:
+            collection
+            for collection in collections
+            if collection["name"] == collection_name
+        ]
+        if preexisting:
+            pre_collection = preexisting[0]
+            # if it has no "indexing", it is a legacy collection
+            pre_col_options = pre_collection.get("options") or {}
+            if "indexing" not in pre_col_options:
+                # legacy collection on DB
+                if requested_indexing_policy == DEFAULT_INDEXING_OPTIONS:
+                    warnings.warn(
+                        (
+                            f"Collection '{collection_name}' is "
+                            "detected as legacy and has indexing turned "
+                            "on for all fields. This implies stricter "
+                            "limitations on the amount of text each entry "
+                            "can store. Consider reindexing anew on a "
+                            "fresh collection to be able to store longer texts."
+                        ),
+                        UserWarning,
+                        stacklevel=2,
+                    )
+                else:
+                    raise ValueError(
+                        f"Collection '{collection_name}' is "
+                        "detected as legacy and has indexing turned "
+                        "on for all fields. This is incompatible with "
+                        "the requested indexing policy for this Vector "
+                        "Store. Consider reindexing anew on a fresh "
+                        "collection with the requested indexing "
+                        "policy, or alternatively leave the indexing "
+                        "settings for this store to their defaults "
+                        "to keep using this collection."
+                    )
+            elif pre_col_options["indexing"] != requested_indexing_policy:
+                # collection on DB has indexing settings, but different
+                options_json = json.dumps(pre_col_options["indexing"])
+                if pre_col_options["indexing"] == DEFAULT_INDEXING_OPTIONS:
+                    default_desc = " (default setting)"
+                else:
+                    default_desc = ""
+                raise ValueError(
+                    f"Collection '{collection_name}' is "
+                    "detected as having the following indexing policy: "
+                    f"{options_json}{default_desc}. This is incompatible "
+                    "with the requested indexing policy for this Vector "
+                    "Store. Consider reindexing anew on a fresh "
+                    "collection with the requested indexing "
+                    "policy, or alternatively align the requested "
+                    "indexing settings to the collection to keep using it."
+                )
+            else:
+                # the discrepancies have to do with other settings than indexing
+                raise
```

Comment:
Which exception is raised ?

Reply:
Ah, good catch! It should be the exception raised in `_[a]provision_collection`. Indeed this handler function is called only within those `except` blocks, and I believe the "last exception in the current scope" ([ref](_validate_create_collection_indexing)) is maintained across function calls.

But I agree: it has become a bit obscure after the refactoring into the separate `_validate_create_collection_indexing`. I will adjust for clarity.

Reply:
Then maybe pass the original exception as argument and re-raise it ?

Reply:
Disregard that, this should be a `return False` to let the caller know _it_ has to reraise.